File: pt6/1242.web-crawler-multithreaded.go
package main

import (
	"net/url"
	"sync"
)

type HtmlParser interface {
	GetUrls(url string) []string
}

func crawl(startUrl string, htmlParser HtmlParser) []string {
	host := crawlerHost(startUrl)
	seen := map[string]bool{startUrl: true}
	var mu sync.Mutex
	var wg sync.WaitGroup

	var visit func(string)
	visit = func(page string) {
		defer wg.Done()
		for _, next := range htmlParser.GetUrls(page) {
			if crawlerHost(next) != host {
				continue
			}

			mu.Lock()
			if !seen[next] {
				seen[next] = true
				wg.Add(1)
				go visit(next)
			}
			mu.Unlock()
		}
	}

	wg.Add(1)
	go visit(startUrl)
	wg.Wait()

	ans := make([]string, 0, len(seen))
	for page := range seen {
		ans = append(ans, page)
	}
	return ans
}

func crawlerHost(raw string) string {
	parsed, _ := url.Parse(raw)
	return parsed.Host
}

/*
Explanation

This is a graph traversal where each HtmlParser.GetUrls call may block, so Go
goroutines are useful. A WaitGroup tracks active crawl tasks, and a mutex
protects the shared seen map.

The critical section is small: check whether a same-host URL is unseen, mark it
seen, and launch exactly one goroutine for it. Marking before launching avoids
duplicate work when multiple pages link to the same URL.

The WaitGroup starts with startUrl. Each discovered URL increments the counter
before its goroutine starts; each goroutine calls Done when finished.

Edge cases: repeated links, cycles, external host links, and arbitrary
completion order. Output order is intentionally unspecified.

Time complexity: O(V + E) over reachable same-host pages, ignoring network
latency.
Space complexity: O(V) for seen plus goroutine stack overhead.
*/

File: pt6/1242.web-crawler-multithreaded.java
import java.net.URI;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/*
interface HtmlParser {
    List<String> getUrls(String url);
}
*/

class Solution {
    public List<String> crawl(String startUrl, HtmlParser htmlParser) {
        String hostname = host(startUrl);
        Set<String> seen = ConcurrentHashMap.newKeySet();
        Queue<Future<List<String>>> futures = new ArrayDeque<>();
        ExecutorService executor = Executors.newFixedThreadPool(16);

        seen.add(startUrl);
        futures.offer(executor.submit(() -> htmlParser.getUrls(startUrl)));

        try {
            while (!futures.isEmpty()) {
                Future<List<String>> future = futures.poll();
                for (String next : future.get()) {
                    if (host(next).equals(hostname) && seen.add(next)) {
                        futures.offer(executor.submit(() -> htmlParser.getUrls(next)));
                    }
                }
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } catch (ExecutionException e) {
            throw new RuntimeException(e);
        } finally {
            executor.shutdown();
        }

        return new ArrayList<>(seen);
    }

    private String host(String rawUrl) {
        return URI.create(rawUrl).getHost();
    }
}

/*
Explanation

This is the graph traversal from the single-threaded crawler, but calls to
HtmlParser.getUrls are submitted to an ExecutorService so multiple blocking URL
fetches can run concurrently.

ConcurrentHashMap.newKeySet() is the key Java data structure: it lets us mark a
URL as seen atomically with seen.add(next), preventing duplicate crawl tasks
when multiple pages link to the same URL. A queue tracks submitted Future
objects whose results still need to be processed.

Edge cases: repeated links, cycles, external host links, and arbitrary task
completion order. Output order is unspecified.

Time complexity: O(V + E) over reachable same-host links, ignoring wait time.
Space complexity: O(V) for seen URLs and pending futures.
*/
